Distribution/Server/Features/Tags.hs

```diff
@@ -195,20 +230,78 @@ tagsFeature CoreFeature{ queryGetPackageIndex
         pkgMap <- liftM packageTags $ queryState tagsState GetPackageTags
         return $ Map.fromDistinctAscList . map (\pkg -> (pkg, Map.findWithDefault Set.empty pkg pkgMap)) $ Set.toList pkgs
 
+    mergeTags :: Tag -> ServerPartE ()
+    mergeTags deprTag = do
+        tags <- optional $ look "tags"
+        index <- queryGetPackageIndex
+        case simpleParse =<< tags of
+            Just (Tag orig) -> do
+                void $ updateState tagsAlias $ AddTagAlias (Tag orig) deprTag
+                void $ constructMergedTagIndex (Tag orig) deprTag index
+            _ -> errBadRequest "Tag not recognised" [MText "Couldn't parse tag. It should be a single tag."]
+
+    -- tags on merging
+    constructMergedTagIndex :: forall m. (Functor m, MonadIO m) => Tag -> Tag -> PackageIndex PkgInfo -> m PackageTags
+    constructMergedTagIndex orig depr = foldM addToTags emptyPackageTags . PackageIndex.allPackagesByName
+      where addToTags calcTags pkgList = do
+                let info = pkgDesc $ last pkgList
+                    !pn = packageName info
+                pkgTags <- queryTagsForPackage pn
+                if Set.member depr pkgTags
+                    then do
+                        let newTags = Set.delete depr (Set.insert orig pkgTags)
+                        void $ updateState tagsState $ SetPackageTags pn newTags
+                        runHook_ tagsUpdated (Set.singleton pn, newTags)
+                        return $ setTags pn newTags calcTags
+                    else return $ setTags pn pkgTags calcTags
+
     putTags :: PackageName -> ServerPartE ()
     putTags pkgname = do
       guardValidPackageName pkgname
-      guardAuthorisedAsMaintainerOrTrustee pkgname
-      mtags <- optional $ look "tags"
-      case simpleParse =<< mtags of
-          Just (TagList tags) -> do
-              calcTags <- fmap (packageToTags pkgname) $ readMemState calculatedTags
-              let tagSet = Set.fromList tags `Set.union` calcTags
-              void $ updateState tagsState $ SetPackageTags pkgname tagSet
-              runHook_ tagsUpdated (Set.singleton pkgname, tagSet)
-              return ()
+      addns <- optional $ look "addns"
+      delns <- optional $ look "delns"
+      raddns <- optional $ look "raddns"
+      rdelns <- optional $ look "rdelns"
```

Comment:
These front-end bits don't really belong here. `putTags` should accept all these things as function args and the html layer should get them from the POST params and do the parsing.